lib/si-split-graph/src/lib.rs

```diff
@@ -0,0 +1,1178 @@
+    R: SubGraphReader<N, E, K>,
+    W: SubGraphWriter<N, E, K>,
+{
+    pub fn new(reader: &'a R, writer: &'b W, split_max: u16) -> Self {
+        let mut first_subgraph = SubGraph::new();
+        let root_id = Ulid::new();
+        let root_index = first_subgraph
+            .graph
+            .add_node(SplitGraphNodeWeight::GraphRoot {
+                id: root_id,
+                merkle_tree_hash: MerkleTreeHash::nil(),
+            });
+        first_subgraph.node_index_by_id.insert(root_id, root_index);
+        first_subgraph.root_index = root_index;
+
+        Self {
+            supergraph: SuperGraph {
+                addresses: vec![],
+                root_index: SplitGraphNodeIndex {
+                    index: root_index,
+                    subgraph: 0,
+                },
+                split_max,
+            },
+            subgraphs: vec![first_subgraph],
+            reader,
+            writer,
+        }
+    }
+
+    pub fn node_count(&self) -> usize {
+        self.subgraphs.iter().fold(0, |count, subgraph| {
+            count.saturating_add(subgraph.node_index_by_id.len())
+        })
+    }
+
+    pub fn root_id(&self) -> SplitGraphResult<SplitGraphNodeId> {
+        self.node_weight_by_index(self.supergraph.root_index)
+            .map(|node| node.id())
+            .ok_or(SplitGraphError::RootNodeNotFound)
+    }
+
+    pub fn subgraph_count(&self) -> usize {
+        self.subgraphs.len()
+    }
+
+    pub async fn new_with_addresses(
+        reader: &'a R,
+        writer: &'b W,
+        addresses: &[SubGraphAddress],
+        split_max: u16,
+    ) -> SplitGraphResult<Self> {
+        // We need to do this without creating a root node
+        let mut split_graph = Self::new(reader, writer, split_max);
+        split_graph.add_subgraphs(addresses).await?;
+
+        Ok(split_graph)
+    }
+
+    pub async fn add_subgraphs(
+        &mut self,
+        subgraph_addresses: &[SubGraphAddress],
+    ) -> SplitGraphResult<()> {
+        self.supergraph.addresses.extend(subgraph_addresses.iter());
+
+        for (idx, address) in self.supergraph.addresses.iter().enumerate() {
+            if self.subgraphs.get(idx).is_none() {
+                let subgraph = self
+                    .reader
+                    .read_subgraph(*address)
+                    .await
+                    .map_err(|err| SplitGraphError::SubGraphRead(*address, err.to_string()))?;
+
+                self.subgraphs.push(subgraph);
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn recalculate_merkle_tree_hashes_based_on_touched_nodes(&mut self) {
+        self.subgraphs
+            .iter_mut()
+            .for_each(|subgraph| subgraph.recalculate_merkle_tree_hash_based_on_touched_nodes());
+    }
+
+    pub fn recalculate_entire_merkle_tree_hashes(&mut self) {
+        self.subgraphs
+            .iter_mut()
+            .for_each(|subgraph| subgraph.recalculate_entire_merkle_tree_hash());
+    }
```

Comment:
These don't appear to propagate the merkle tree hash across subgraphs? The subgraph recalculations appear to do a dfs post-order, but cross subgraph edges don't appear to affect the other subgraph?

Reply:
correct. The merkle tree hashes are restricted to each subgraph, and updates apply only to the subgraph. If a child of a parent in another subgraph changes, it doesn't matter. The merkle tree for the subgraph that changed will be updated, and we will apply the update to that hash. The change will be propagated up to the "supergraph" because the hash for that subgraph will change, so we will get a new graph in the end.